rust/connlib/libs/client/src/lib.rs

```diff
@@ -1,25 +1,246 @@
+                    }
+                }
+            });
+
+            let control_signaler = ControlSignaler { control_signal: connection.sender_with_topic("client".to_owned()) };
+            let tunnel = fatal_error!(
+                Tunnel::new(private_key, control_signaler.clone(), callbacks.clone()).await,
+                runtime_stopper,
+                &callbacks
+            );
+
+            let mut control_plane = ControlPlane {
+                tunnel: Arc::new(tunnel),
+                control_signaler,
+                tunnel_init: Mutex::new(false),
+            };
+
+            tokio::spawn(async move {
+                let mut interval = tokio::time::interval(Duration::from_secs(10));
+                loop {
+                    tokio::select! {
+                        Some((msg, reference)) = control_plane_receiver.recv() => {
+                            match msg {
+                                Ok(msg) => control_plane.handle_message(msg, reference).await?,
+                                Err(err) => control_plane.handle_error(err, reference).await,
+                            }
+                        },
+                        _ = interval.tick() => control_plane.stats_event().await,
+                        else => break
+                    }
+                }
+                Result::Ok(())
+            });
+
+            tokio::spawn(async move {
+                let mut exponential_backoff = ExponentialBackoffBuilder::default().build();
+                loop {
+                    // `connection.start` calls the callback only after connecting
+                    tracing::debug!("Attempting connection to portal...");
+                    let result = connection.start(vec!["client".to_owned()], || exponential_backoff.reset()).await;
+                    tracing::warn!("Disconnected from the portal");
+                    if let Err(e) = &result {
+                        tracing::warn!(error = ?e, "Portal connection error");
+                    }
+                    if let Some(t) = exponential_backoff.next_backoff() {
+                        tracing::warn!("Error connecting to portal, retrying in {} seconds", t.as_secs());
+                        let _ = callbacks.on_error(&result.err().unwrap_or(Error::PortalConnectionError(tokio_tungstenite::tungstenite::Error::ConnectionClosed)));
+                        tokio::time::sleep(t).await;
+                    } else {
+                        tracing::error!("Connection to portal failed, giving up");
+                        fatal_error!(
+                            result.and(Err(Error::PortalConnectionError(tokio_tungstenite::tungstenite::Error::ConnectionClosed))),
+                            runtime_stopper,
+                            &callbacks
+                        );
+                    }
+                }
+
+            });
```

Comment:
Even for the client, I am wondering whether it wouldn't be cleaner to just re-initialize the `PhoenixChannel` on disconnect.

Reply:
As it is structured now I don't think it changes too much, this re-starts the websocket anyways.